### PR TITLE
VBOverloadResolutionPriority: break attribute binding cycles, clone a bunch of C# tests

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @333fred, @dotnet/roslyn-compiler Please review

**Review by 333fred — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler For the second review

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/OverloadResolutionPriorityTests.vb [hunk 1]
```diff
@@ -1560,5 +1562,1613 @@ End Module
...
+    System.Console.WriteLine(1)
+End Sub
+"
+
+            Dim i2Source = "
+public Shared Sub M(x As I2)
+    throw DirectCast(Nothing, System.Exception)
+End Sub
+"
+
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/Semantics/OverloadResolutionPriorityTests.vb:1615]: > It looks like these are not used.

Looks this way. Will remove.

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/OverloadResolutionPriorityTests.vb [hunk 2]
```diff
@@ -1560,5 +1562,1613 @@ End Module
...
+    Shared Sub Main
+        Dim d = new Derived()
+        d.M(""test"")
+    End Sub
+End Class
+"
+
+            Dim compilation = CreateCompilation({source, OverloadResolutionPriorityAttributeDefinitionVB}, options:=TestOptions.DebugExe)
+
+            ' Note, VB silently ignores attributes at wrong location
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/OverloadResolutionPriorityTests.vb:2066]: >Note, VB silently ignores attributes at wrong location [](http://example.com/codeflow?start=14&length=54)

Do we need a test for C# consuming such cases?

Comment (cston) [reply src/Compilers/VisualBasic/Test/Semantic/Semantics/OverloadResolutionPriorityTests.vb:2066]: It looks like there is such a test already: `Overrides_ChangePriorityInMetadata_Methods`.

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/OverloadResolutionPriorityTests.vb [hunk 3]
```diff
@@ -1560,5 +1562,1613 @@ End Module
...
+
+    Sub Main
+        Dim i3 As I3 = Nothing
+        i3.M()
+    End Sub
+End Module
+"
+
+            Dim compilation = CreateCompilation(source, options:=TestOptions.DebugExe)
+            compilation.AssertTheseEmitDiagnostics(
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/OverloadResolutionPriorityTests.vb:2588]: >compilation.AssertTheseEmitDiagnostics [](http://example.com/codeflow?start=12&length=38)

The equivalent C# test avoids verifying diagnostics, presumably to increase the chance of a cycle in attribute binding when using the semantic model. Is that not useful here?

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/Semantics/OverloadResolutionPriorityTests.vb:2588]: >Is that not useful here?

It is done below. The compilation is recreated after we check diagnostics

**Review by cston — APPROVED** [review]

**Pull request merged**